src/main/java/org/commonjava/indy/service/archive/controller/ArchiveController.java

```diff
@@ -0,0 +1,256 @@
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+@ApplicationScoped
+public class ArchiveController {
+
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    @Inject
+    PreSeedConfig preSeedConfig;
+
+    @Inject
+    ObjectMapper objectMapper;
+
+    private ExecutorService executorService;
+
+    private CloseableHttpClient client;
+
+    private String contentDir;
+
+    private String archiveDir;
+
+    private final String CONTENT_DIR = "/content";
+
+    private final String ARCHIVE_DIR = "/archive";
+
+    @PostConstruct
+    public void init()
+    {
+        int threads = 4 * Runtime.getRuntime().availableProcessors();
+        executorService = Executors.newFixedThreadPool( threads, ( final Runnable r ) -> {
+            final Thread t = new Thread( r );
+            t.setDaemon( true );
+            return t;
+        } );
+
+        final PoolingHttpClientConnectionManager ccm = new PoolingHttpClientConnectionManager();
+        ccm.setMaxTotal( 500 );
+
+        RequestConfig rc = RequestConfig.custom().build();
+        final HttpClientBuilder builder = HttpClients.custom().setConnectionManager( ccm ).setDefaultRequestConfig( rc );
+        client = builder.build();
+
+        String storeDir = preSeedConfig.storageDir.orElse( "data" );
+        contentDir = String.format( "%s%s", storeDir, CONTENT_DIR );
+        archiveDir = String.format( "%s%s", storeDir, ARCHIVE_DIR );
+    }
+
+    @PreDestroy
+    public void destroy()
+    {
+        IOUtils.closeQuietly( client, null );
+    }
+
+    public void downloadArtifacts( final Map<String, String> downloadPaths, final HistoricalContentDTO content )
+    {
+        BasicCookieStore cookieStore = new BasicCookieStore();
+        ExecutorCompletionService executor = new ExecutorCompletionService<>( executorService );
+
+        String contentBuildDir = String.format( "%s/%s", contentDir, content.getBuildConfigId() );
+        File dir = new File( contentBuildDir );
+        dir.deleteOnExit();
+
+        fileTrackedContent( contentBuildDir, content );
+
+        for ( String path : downloadPaths.keySet() )
+        {
+            String filePath = downloadPaths.get( path );
+            executor.submit( download( contentBuildDir, path, filePath, cookieStore ) );
+        }
+    }
+
+    public File generateArchiveZip( final HistoricalContentDTO content )
+            throws IOException
+    {
+        String contentBuildDir = String.format( "%s/%s", contentDir, content.getBuildConfigId() );
+        File dir = new File( contentBuildDir );
+        if ( !dir.exists() )
+        {
+            return null;
```

Comment:
If this method will possibly return null, I recommend to return an Optional<File> here and let the caller to decide how to handle when Optional.isEmpty(). This will reduce the NPE from bottom